api/src/main/java/org/infinispan/api/reactive/KeyValueStore.java

```diff
@@ -0,0 +1,131 @@
+package org.infinispan.api.reactive;
+
+import java.util.Map;
+import java.util.concurrent.CompletionStage;
+
+import org.infinispan.api.Experimental;
+import org.infinispan.api.reactive.listener.KeyValueStoreListener;
+import org.infinispan.api.reactive.query.QueryRequest;
+import org.reactivestreams.Publisher;
+
+/**
+ * A Reactive Key Value Store provides a highly concurrent and distributed data structure, non blocking and using
+ * reactive streams.
+ * <p>
+ *
+ * </p>
+ *
+ * @author Katia Aresti, karesti@redhat.com
+ * @see <a href="http://www.infinispan.org">Infinispan documentation</a>
+ * @since 10.0
+ */
+@Experimental
+public interface KeyValueStore<K, V> {
+
+   /**
+    * Get the value of the Key if such exists
+    * @param key
+    * @return the value
+    */
+   CompletionStage<V> get(K key);
+
+   /**
+    * Insert the key/value if such key does not exist
+    *
+    * @param key
+    * @param value
+    * @return Void
+    */
+   CompletionStage<Void> insert(K key, V value);
```

Comment:
how do I know if the `insert()` succeeds or not?

Reply:
we don't know with the actual API. Should we consider a WriteResult ?

Reply:
I was thinking about `CompletionStage<Boolean>` or similar

Reply:
If it doesn't succeed because of an error, then it should completeExceptionally. Otherwise, why would it fail ?

Reply:
@karesti I'm ok with either way

Reply:
@tristantarrant If two concurrent inserts happen, one might insert and the other not?

Reply:
In a caching use case (and independent of having a cache API, which is being worked on), an insert could happen as a result of two things (assuming no lock timeouts or other exceptions):

1. Insert some data (key=a,value=b) into key/value store as a result of reading it from X. If N threads read the same thing from X and call insert, one of them would win and the others not. Some APIs *might * want to know which one won. To the end result, the return of insert does not matter, the cache has a/b pairing.

2. Two threads attempt to insert new data into X, which in turn stores in the key/value store. Say key=a and thread 1 stores value=b and thread-2 stores value=c. Assuming no lock timeouts, how would the user know which of the inserts won? You'd have to check the cache contents with a get?

Reply:
Having said that, I do like the idea of starting with the least powerful option (one that returns CS wrapping Void), and then add a more powerful if we *really* need it (ione that returns CS wrapping a Boolean). I just want to make sure we do think about this with care.

Reply:
I added a Boolean but at this point we can change everything